### PR TITLE
HMR plugins

### PR DESCRIPTION
**Pull request by basarat**
⚠️ Work in progress. Will message once done :rose: 

All HMR issues seem to be about not wanting the cache to clear for certain modules:

* https://github.com/fuse-box/fuse-box/issues/148 (`<Provider> does not support changingstoreon the fly...`)
* https://github.com/fuse-box/fuse-box/issues/115 (user wants to exclude the main module)
* https://github.com/fuse-box/fuse-box/issues/98 (Meta issue but only links to #115 right now)

### CODE REVIEW

**Comment by basarat** [comment]
Don't merge. Will talk more tomorrow 🌹

**Comment by nchanged** [comment]
Okay!

**Comment by leon-andria** [comment]
Does HMR should be multiple in an application ?
Or they should be chainable ?
It might dangerous to make it just an array

**Comment by nchanged** [comment]
It might, not merging anything now, but we need to have HMR plugin support.

**Comment by leon-andria** [comment]
Of course. HMR is very important.
I just highlight the fact that speed implementation does match with quality and sustainability.  :-)

**Comment by basarat** [comment]
@nchanged just a quick review (don't merge as I need to test it). Main commit is https://github.com/fuse-box/fuse-box/pull/158/commits/3e87d418b48e77602183aad744499b75ab5508bb

The way it works: 

* User passes option `hmrStatefulPlugins` to `devServer` and we just send it to `HotModuleReloadPlugin` : https://github.com/fuse-box/fuse-box/pull/158/commits/3e87d418b48e77602183aad744499b75ab5508bb#diff-8c475be8a75f3fc4045c0fc95d5c54dfR61 
*  That in turn just passes it to the front-end `fusebox-hot-reload` : https://github.com/fuse-box/fuse-box/pull/158/commits/3e87d418b48e77602183aad744499b75ab5508bb#diff-b7e42c8e1abbd7f575cbbe589fce3215R8 
* which then uses it to register a plugin (with essentially itself) https://github.com/fuse-box/fuse-box/pull/158/commits/3e87d418b48e77602183aad744499b75ab5508bb#diff-b7e42c8e1abbd7f575cbbe589fce3215R17 for a different behavior (flushing other files except the stateful ones unless some stateful one has changed in which case all files are flushed). 

My suggestion: 

* I remove `addHMRPlugin` and just move the code inline here : https://github.com/fuse-box/fuse-box/pull/158/commits/3e87d418b48e77602183aad744499b75ab5508bb#diff-b7e42c8e1abbd7f575cbbe589fce3215R46
* Or I can still leave the code for `addHMRPlugin` and just not use it. 

As always, thanks! :rose:

**Comment by nchanged** [comment]
That's just terrific! thanks!! Just woke up and what a nice surprise!

**Comment by nchanged** [comment]
Would be cool to have one flush method with options. Maybe regexp... or test filter function?! :)

**Comment by basarat** [comment]
> Would be cool to have one flush method with options. Maybe regexp... or test filter function?! :)

Done! :rose: https://github.com/fuse-box/fuse-box/pull/158/commits/69395fcdb03bce94f5372908652a811a4bfc2a32 

My questions in https://github.com/fuse-box/fuse-box/pull/158#issuecomment-279922333 still stand :heart:

**Comment by nchanged** [comment]
Ok. what about a default `hmr-default-plugin` in [here](https://github.com/fuse-box/fuse-box/tree/master/modules) ? And I think you missed the CSS reload. We need tests for that...

**Comment by basarat** [comment]
> And I think you missed the CSS reload. We need tests for that...

Explicitly didn't cater for it https://github.com/fuse-box/fuse-box/pull/158/commits/3e87d418b48e77602183aad744499b75ab5508bb#diff-b7e42c8e1abbd7f575cbbe589fce3215R35 I don't see people wanting to preventing the default behavior ( __fsbx_css https://github.com/fuse-box/fuse-box/pull/158/commits/3e87d418b48e77602183aad744499b75ab5508bb#diff-b7e42c8e1abbd7f575cbbe589fce3215R60  ) for CSS files 

> what about a default hmr-default-plugin in here 

Don't understand. I'm available on gitter for the next 1 hour https://gitter.im/basarat if you want to discuss there :heart:

**Comment by nchanged** [comment]
Wow! Is it ready?)

**Comment by basarat** [comment]
Nope. Working on adding devserver support for loading these 🌹

File: docs/loader-api.md [hunk 1]
```diff
@@ -141,3 +141,60 @@ require("~/lib/utils")
...
+    /** 
+     * If true is returned by the plugin
+     *  it means that module change has been handled
+     *  by plugin and no special work is needed by FuseBox
+     **/
+    hmrUpdate?(evt: SourceChangedEvent): boolean;
+}
+
+/** Where */
+type SourceChangedEvent = {
```
Comment (Thanood) [inline docs/loader-api.md]: You've changed `type` to `interface` in `LoaderAPI.ts` - I'd propose to change this line, too. 😃

Comment (basarat) [inline docs/loader-api.md]: I'm scared of people being confused by which interface they need to implement. I know sounds weird but who knows 🌹

**Comment by nchanged** [comment]
Hi! You should probably sync with master =)

**Comment by basarat** [comment]
Will do :rose:

**Comment by nchanged** [comment]
Interesting.. it's not happening in master...

**Comment by basarat** [comment]
> Interesting.. it's not happening in master...

Suspect me touching the loader API has pushed the size over the hard coded limit `5300` ?

**Comment by nchanged** [comment]
Okay, can you increase it a little? But just enough=)

**Comment by basarat** [comment]
@nchanged Ready to merge / release whenever you want :1st_place_medal:  no hurry :heart:

magic

Tested on the fuse-box-reflux example. Thanks for making such an awesome ecosystem! :rose: :heart:

**Comment by nchanged** [comment]
@basarat :fire: That's just astonishing!

**Comment by nchanged** [comment]
Coming in the next release!

**Comment by basarat** [comment]
Note to users : Docs for this feature are published as well : http://fuse-box.org/#loader-plugins :rose:

**Comment by tomitrescak** [comment]
Thanks for the amazing PR! Looking forward to it like a baby to a new rattle ;)
Will you provide an example of how to include React HMR in an app? I looked at docs but to be hones, not too smart from it :/

**Comment by basarat** [comment]
# Once this is released 

@tomitrescak After this is released you would copy paste this (from the docs http://fuse-box.org/#loader-plugins)

```
const registerStatefulModules = (moduleNames: string[]) => FuseBox.addPlugin({
  hmrUpdate: ({ type, path, content }) => {
    if (type === "js") {
      const isModuleNameInPath = (path) => moduleNames.some(name => path.includes(name));

      /** If a stateful module has changed reload the window */
      if (isModuleNameInPath(path)) {
        window.location.reload();
      }

      /** Otherwise flush the other modules */
      FuseBox.flush(function(fileName) {
        return !isModuleNameInPath(fileName);
      });
      /** Patch the module at give path */
      FuseBox.dynamic(path, content);

      /** Re-import / run the mainFile */
      if (FuseBox.mainFile) {
        FuseBox.import(FuseBox.mainFile)
      }

      /** We don't want the default behavior */
      return true;
    }
  }
});

registerStatefulModules(['TestStore', 'actions/index']);
```
into the bottom of this file: https://github.com/fuse-box/fuse-box-ts-react-reflux-seed/blob/0c659e56eca78f8edf8f673c7b55e81385e94043/src/client/store/TestStore.ts 


# Future

Ideally this code should be written in something that is only loaded by the dev server instead of polluting your application.  I was thinking 

```ts
.devServer('>yourApp.tsx', {serverBundle:'>moreCodeThatCustomizesTheServer'})
```

Haven't finalized that idea yet though :rose:

**Comment by nchanged** [comment]
@basarat can this file be at the projects' root? With test for `process.env.NODE_ENV` that would make sense =)

**Comment by basarat** [comment]
> can this file be at the projects' root?

Not quite as the project root is run again and again by 

```ts
/** Re-import / run the mainFile */
      if (FuseBox.mainFile) {
        FuseBox.import(FuseBox.mainFile)
      }
```

So it would register the plugin again and again. Has to be in some stateful file and since we reload the page on a stateful file change its not a problem. Ideally it would be in a server only bundle making it immune to HMR updates :rose:

**Comment by nchanged** [comment]
@basarat but we could blacklist it :-)
Besides people will use to reload their modules without touching an entry point. We just need to make it simple to use. I gave you write access to this repo https://github.com/fuse-box/fuse-box-hmr-examples could put cook some examples for people? ;-)

**Comment by tomitrescak** [comment]
Examples would be great! I'm still not sure what is expected in registerStatefulModules. Do I need to register every single ui component, or just a router, for example. 

So an example of SPA, with (react/mobx/yester) router and couple routes would be peachy;)

**Comment by basarat** [comment]
@tomitrescak https://github.com/fuse-box/fuse-box-ts-react-reflux-seed has been updated. `improveHMR` 👍 

Will work on a yester example next :rose:

**Comment by basarat** [comment]
> Do I need to register every single ui component, or just a router, for example.

No. Just the files containing `state` and the `router`. Hold tight. Example coming :heart:

**Comment by tomitrescak** [comment]
In the seed example , why did you use 'TestStore' module as path when you used 'actions/index's? It seems that correct path would be 'store/TestStore', what am I missing?

**Comment by basarat** [comment]
@tomitrescak as long as *some* portion of the path matches it is okay. https://github.com/fuse-box/fuse-box-ts-react-reflux-seed/blob/df23675337e37013b6d8bb9803de0423400b029a/src/client/hmr.ts#L5 

I should just make it `store` and `actions` :rose:

**Pull request merged**